test/balance_reader.spec.ts

```diff
@@ -0,0 +1,171 @@
+import BN from "bn.js";
+
+import {closeAuction, applyBalances} from "./utilities";
+import {getBalanceState, getWithdrawableAmount} from "../src/balance_reader";
+
+const MockContract = artifacts.require("MockContract");
+const BatchExchange = artifacts.require("BatchExchange");
+
+contract("BatchExchange utils", async (accounts) => {
+  describe("getBalanceState()", async () => {
+    it("retrieves balance as in storage", async () => {
+      const erc20 = await MockContract.new();
+      await erc20.givenAnyReturnBool(true);
+      const batchExchange = await BatchExchange.new(1, erc20.address);
+      await closeAuction(batchExchange);
+
+      // amount is hex 100 to catch possible zero padding issues in the function
+      await batchExchange.deposit(erc20.address, 0x100);
+      await applyBalances(accounts[0], batchExchange, [erc20.address]);
+
+      const balance = await getBalanceState(
+        accounts[0],
+        erc20.address,
+        batchExchange.address,
+        web3
+      );
+      assert.equal(balance.toNumber(), 0x100);
+    });
+  });
+
+  describe("getWithdrawableAmount()", async () => {
+    const startingAmount = new BN("100");
+
+    it("returns zero for future withdraw request", async () => {
+      const erc20 = await MockContract.new();
+      await erc20.givenAnyReturnBool(true);
+      const batchExchange = await BatchExchange.new(1, erc20.address);
+      await closeAuction(batchExchange);
+      await batchExchange.deposit(erc20.address, startingAmount);
+      await applyBalances(accounts[0], batchExchange, [erc20.address]);
+
+      const futureBatchId = 2 ** 32 - 1;
+      await batchExchange.requestFutureWithdraw(
+        erc20.address,
+        startingAmount,
+        futureBatchId
+      );
+      const withdrawableAmount = await getWithdrawableAmount(
+        accounts[0],
+        erc20.address,
+        batchExchange,
+        web3
+      );
+      assert.equal(withdrawableAmount.toString(), "0");
+    });
+    it("returns amount when withdraw was already requested", async () => {
+      const erc20 = await MockContract.new();
+      await erc20.givenAnyReturnBool(true);
+      const batchExchange = await BatchExchange.new(1, erc20.address);
+      await closeAuction(batchExchange);
+      await batchExchange.deposit(erc20.address, startingAmount);
+      await applyBalances(accounts[0], batchExchange, [erc20.address]);
+
+      await batchExchange.requestWithdraw(erc20.address, startingAmount);
+      await closeAuction(batchExchange);
+      const withdrawableAmount = await getWithdrawableAmount(
+        accounts[0],
+        erc20.address,
+        batchExchange,
+        web3
+      );
+      assert.equal(withdrawableAmount.toString(), startingAmount.toString());
+      const transcript = await batchExchange.withdraw(
+        accounts[0],
+        erc20.address
+      );
+      assert.equal(
+        withdrawableAmount.toString(),
+        transcript.logs[0].args.amount.toString()
+      );
```

Comment:
ubernit: If you wanted to reduce the tests a bit, I'm not sure the withdraw has to actually be tested here, but feel free to leave.

Reply:
I'd keep it there, I think it makes the purpose of the function clearer. I was even considering testing if `withdraw` reverts in the case when the output of the tested function is zero but decided against it.